File: co_testing/unit_tests/QMF_Analysis/test_QMFA.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  P & D Embedded Systems and Multimedia [H09M0a] 2015-2016
%  Subband-Coding 
%
%  Mohit Dandekar
%  John O'Callaghan
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Unit Test for QMF analysis bank
%
%  + produces random data of length N
%  + Processes using QMF_analysis MATLAB kernel to produce output
%  + Writes the testvectors: input_data, filter, expected outputs
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

clear all;
close all;
%clc;

if(exist('OCTAVE_VERSION', 'builtin') ~= 0)
  pkg load signal;
end
addpath('../../../MATLAB/src');
addpath('../../../MATLAB/utils');


f2 = fopen('./test_vectors/test_name.txt','w');
fprintf(f2,'QMF Analysis');
fclose(f2);

fprintf('#################################################################\n');
printtime();
fprintf('     Unit Testcase for QMF analysis filter bank                  \n\n');




load('filters_40tap.mat');
load('../../../training_data/words_m')
h0 = convert_to_FXPT(h0,16,15);
f0 = convert_to_FXPT(f0,16,15);

%yl = convert_to_FXPT((rand(N,1)-0.5),16,15);
%yr = convert_to_FXPT((rand(N,1)-0.5),16,15);
xl = x(1:2:length(x));
yl = resample(xl,8000,16000); % includes anti-alising
yl = yl(1:(floor(length(yl)/10/24))*24);
xr = x(2:2:length(x));
yr = resample(xr,8000,16000); % includes anti-alising
yr = yr(1:(floor(length(yr)/10/24))*24);

yl = convert_to_FXPT(yl,16,15);
yr = convert_to_FXPT(yr,16,15);

M = 40;
L = 24;
N = length(yl);
P = N/24;
fprintf('Number of Input blocks = %d  Block Size = %d\n',P,L);

fs = fs/2;



mode = 0;

history1l = zeros(N/2/P+M-1,2);
history1r = zeros(N/2/P+M-1,2);
start_indx = [1 1 1 1]; end_indx = [N/P N/2/P N/4/P N/8/P];
incr = end_indx;

fprintf('Analysis Filter Bank\n');fprintf('Block : [%d]',P);
for i=1:P

fprintf('%8d',i);%fflush(1);

[bank0l(start_indx(2):end_indx(2),1), bank1l(start_indx(2):end_indx(2),1), history1l] = QMF_analysis(yl(start_indx(1):end_indx(1),1), h0, history1l,mode);

[bank0r(start_indx(2):end_indx(2),1), bank1r(start_indx(2):end_indx(2),1), history1r] = QMF_analysis(yr(start_indx(1):end_indx(1),1), h0, history1r,mode);

fprintf('\b\b\b\b\b\b\b\b');
start_indx = start_indx + incr;
end_indx = end_indx + incr;
end

fprintf('\nWriting Testvectors\n');

f0 = fopen('./test_vectors/filter.txt','w');
for i=1:M
  fprintf(f0,'%d\n',h0(i));
end
fclose(f0);

f1 = fopen('./test_vectors/data_in.txt','w');
for i=1:N
  fprintf(f1,'%d,%d\n',yl(i),yr(i));
end
fclose(f1);

f2 = fopen('./test_vectors/bank0_expected.txt','w');
for i=1:N/2
  fprintf(f2,'%d,%d\n',bank0l(i),bank0r(i));
end
fclose(f2);

f3 = fopen('./test_vectors/bank1_expected.txt','w');
for i=1:N/2
  fprintf(f3,'%d,%d\n',bank1l(i),bank1r(i));
end
fclose(f3);

File: co_testing/unit_tests/QMF_Analysis/test_QMFA.c
/*#########################################################################
#  P & D Embedded Systems and Multimedia [H09M0a]
#  Subband-Coding 
#  
#  Mohit Dandekar
#  John O'Callaghan
#########################################################################
#
#  Test Wrapper for unit test QMF analysis filter bank
#
#  + Reads testvectors: data_in, filter, expected bank 0,1 outputs
#  + Calls C-kernel 'QMFA_process' 
#  + compares DUT output with REF (1 bit error allowed)
#######################################################################*/
#include <stdio.h>
#include <stdlib.h>

#include "QMF.h"

#define P 738
#define N P*BLOCK_SIZE 

QMFA_bank_t QMFA_bank;

aud_data_t data_in[P*BLOCK_SIZE];
aud_data_t data_out_bank0_ref[P*BLOCK_SIZE/2];
aud_data_t data_out_bank0_dut[P*BLOCK_SIZE/2];
aud_data_t data_out_bank1_ref[P*BLOCK_SIZE/2];
aud_data_t data_out_bank1_dut[P*BLOCK_SIZE/2];

sint16 filter[FILT_LEN];

aud_data_t temp_buffer0[BLOCK_SIZE];
aud_data_t temp_buffer1[BLOCK_SIZE];

int main(void)
{
  sint32 ret = 0;
  sint32 i,j;

  /*Read Testvector data*/
  printf("Reading Testvectors\n");
  FILE *fp;
  fp = fopen("./test_vectors/data_in.txt","r");
  for(i=0;i<BLOCK_SIZE*P;++i)
  {
    int templ,tempr;
    fscanf(fp,"%d,%d",&templ,&tempr);
    data_in[i].l = templ;
    data_in[i].r = tempr;
  }
  fclose(fp);

  fp = fopen("./test_vectors/filter.txt","r");
  for(i=0;i<40;++i)
  {
    int temp;
    fscanf(fp,"%d",&temp);
    filter[i] = temp;
  }
  fclose(fp);

  fp = fopen("./test_vectors/bank0_expected.txt","r");
  for(i=0;i<BLOCK_SIZE*P/2;++i)
  {
    int templ,tempr;
    fscanf(fp,"%d,%d",&templ,&tempr);
    data_out_bank0_ref[i].l = templ;
    data_out_bank0_dut[i].l = 0;
    data_out_bank0_ref[i].r = tempr;
    data_out_bank0_dut[i].r = 0;
  }
  fclose(fp);

  fp = fopen("./test_vectors/bank1_expected.txt","r");
  for(i=0;i<BLOCK_SIZE*P/2;++i)
  {
    int templ,tempr;
    fscanf(fp,"%d,%d",&templ,&tempr);
    data_out_bank1_ref[i].l = templ;
    data_out_bank1_dut[i].l = 0;
    data_out_bank1_ref[i].r = tempr;
    data_out_bank1_dut[i].r = 0;
  }
  fclose(fp);


  /*Initialization*/
  printf("Initialize: C kernel under Test\n");
  QMFA_bank_init( &QMFA_bank,
                  temp_buffer0,
                  temp_buffer1,
                  filter,
                  BLOCK_SIZE);

  /*Process Call*/
  printf("Process Call: C kernel under Test\n");
  for(i=0;i<P;++i)
  {
    /*Write  Input Buffer of the QMF*/
    for(j=0;j<BLOCK_SIZE;++j)
    {
      QMFA_bank.data_in_buffer[j] = data_in[i*BLOCK_SIZE + j];
    }

    /*QMF process call*/
    QMFA_process( &QMFA_bank);

    /*Read from temporary out buffers to main memory*/
    for(j=0;j<BLOCK_SIZE/2;++j)
    {
       data_out_bank0_dut[i*BLOCK_SIZE/2 + j] = temp_buffer0[j];
       data_out_bank1_dut[i*BLOCK_SIZE/2 + j] = temp_buffer1[j];
    }
  }


  /* Compare Output vs Ref*/
  ret = 0;
  printf("Checking Bank0 output\n");
  for(i=0;i<P*BLOCK_SIZE/2;++i)
  {
    /*Allowing 2 bit errors*/
    if(abs(data_out_bank0_ref[i].l - data_out_bank0_dut[i].l)>1 || abs(data_out_bank0_ref[i].l - data_out_bank0_dut[i].l)>1)
    {
      printf("ERROR : indx = %8d expected = %8d,%8d  output %8d %8d\n",i,data_out_bank0_ref[i].l, 
                                                                          data_out_bank0_ref[i].r,  
                                                                          data_out_bank0_dut[i].l,
                                                                          data_out_bank0_dut[i].r);
      ret += 1;
    }
  }


  printf("Checking Bank1 output\n");
  for(i=0;i<P*BLOCK_SIZE/2;++i)
  {
    /*Allowing 2 bit errors*/
    if(abs(data_out_bank1_ref[i].l - data_out_bank1_dut[i].l)>0 || abs(data_out_bank1_ref[i].l - data_out_bank1_dut[i].l)>0)
    {
      printf("ERROR : indx = %8d expected = %8d,%8d  output %8d %8d\n",i,data_out_bank1_ref[i].l, 
                                                                          data_out_bank1_ref[i].r,  
                                                                          data_out_bank1_dut[i].l,
                                                                          data_out_bank1_dut[i].r);
      ret += 1;
    }
  }


  if(ret)
  {
         printf("                          TEST FAIL                               \n");
         printf("  total errors: %d \n",ret                                           );
         printf("##################################################################\n");

  }else{
         printf("                          TEST PASS                               \n");
         printf("##################################################################\n");
       }

  return 0;
}


